File: test/technology/mol_qca/test_mol_qca_layout.cpp
/*
 * Copyright (c) 2018 - 2023 Marcel Walter
 * Copyright (c) 2023 - present Chair for Design Automation, Technical University of Munich
 * All rights reserved.
 *
 * SPDX-License-Identifier: MIT
 *
 * Licensed under the MIT License
 */

/**
 * @file
 * @brief Tests for `fiction/technology/mol_qca/layout.hpp`.
 * @author Marcel Walter (marcelwa)
 */

#include <catch2/catch_test_macros.hpp>

#include <fiction/technology/mol_qca/layout.hpp>

using namespace fiction;
using namespace fiction::mol_qca;

TEST_CASE("molQCA cell types carry their clock phase", "[mol-qca-layout]")
{
    CHECK(clock_number(cell_type::NORMAL1) == 0);
    CHECK(clock_number(cell_type::NORMAL2) == 1);
    CHECK(clock_number(cell_type::NORMAL3) == 2);
    CHECK(clock_number(cell_type::NORMAL4) == 3);
    CHECK(clock_number(cell_type::INPUT) == 0);

    CHECK(is_normal(cell_type::NORMAL3));
    CHECK(!is_normal(cell_type::OUTPUT));
    CHECK(is_constant(cell_type::CONST_1));
    CHECK(!is_constant(cell_type::NORMAL1));
}

TEST_CASE("molQCA layouts are planar", "[mol-qca-layout]")
{
    layout lyt{{9, 9, 1}, "wire"};

    CHECK(lyt.z() == 0);
    CHECK(lyt.get_layout_name() == "wire");

    lyt.assign_cell_type({0, 4}, cell_type::INPUT);
    lyt.assign_cell_type({1, 4}, cell_type::NORMAL2);
    lyt.assign_cell_type({2, 4}, cell_type::OUTPUT);

    CHECK(lyt.num_cells() == 3);
    CHECK(lyt.num_pis() == 1);
    CHECK(lyt.num_pos() == 1);
    CHECK(clock_number(lyt.get_cell_type({1, 4})) == 1);

    auto copy = lyt;
    copy.assign_cell_type({1, 4}, cell_type::EMPTY);
    CHECK(lyt.get_cell_type({1, 4}) == cell_type::NORMAL2);
    CHECK(copy != lyt);
}

File: test/python/mol_qca/test_mol_qca_layout.py
# Copyright (c) 2018 - 2023 Marcel Walter
# Copyright (c) 2023 - present Chair for Design Automation, Technical University of Munich
# All rights reserved.
#
# SPDX-License-Identifier: MIT
#
# Licensed under the MIT License

from __future__ import annotations

import copy
import re

from mnt.pyfiction.layouts.coords import offset_coordinate
from mnt.pyfiction.mol_qca import mol_qca_cell_type, mol_qca_clock_number, mol_qca_layout


def test_cell_types_carry_clock_phases() -> None:
    assert mol_qca_clock_number(mol_qca_cell_type.NORMAL1) == 0
    assert mol_qca_clock_number(mol_qca_cell_type.NORMAL4) == 3
    assert mol_qca_clock_number(mol_qca_cell_type.INPUT) == 0


def test_planar_layout() -> None:
    layout = mol_qca_layout(offset_coordinate(3, 0, 1), "wire")

    assert layout.z() == 0
    assert layout.get_layout_name() == "wire"

    layout.assign_cell_type(offset_coordinate(0, 0), mol_qca_cell_type.INPUT)
    layout.assign_cell_type(offset_coordinate(1, 0), mol_qca_cell_type.NORMAL1)
    layout.assign_cell_type(offset_coordinate(2, 0), mol_qca_cell_type.NORMAL2)
    layout.assign_cell_type(offset_coordinate(3, 0), mol_qca_cell_type.OUTPUT)

    assert layout.num_cells() == 4
    assert layout.num_pis() == 1
    # the representation colors inputs and outputs with ANSI escapes
    assert re.sub(r"\x1b\[[0-9;]*m", "", repr(layout)) == "iabo\n\n"

    duplicate = copy.deepcopy(layout)
    duplicate.assign_cell_type(offset_coordinate(1, 0), mol_qca_cell_type.EMPTY)
    assert layout.get_cell_type(offset_coordinate(1, 0)) == mol_qca_cell_type.NORMAL1
    assert duplicate != layout
